### PR TITLE
Update redis client

### PR DESCRIPTION
**Pull request by urso**
Rewrite redis client based on outputs.mode package

- support for TLS
- support for SOCKS5
- error handling and retry logic + support for 'Guaranteed' output
- support for failover and loadbalancing if multiple hosts are configured
- support for multiple workers (loadbalancing mode) per redis host
- backoff strategy is redis is unresponsive
- split topology handling from outputer (dedicated config options in redis module)
- Fix response handling draining complete response buffer after sending batch

### CODE REVIEW

File: libbeat/docs/outputconfig.asciidoc [hunk 1]
```diff
@@ -514,68 +511,129 @@ output:
...
+The list of Redis server to connect to. If load balancing is enabled events are
+distributed to these servers. If one server becomes unreachable, the events are
+distributed to the reachable servers only. Each Redis server can be defined as
+`HOST` or `HOST:PORT`. For example: "192.15.3.2", "test.redis.io:12345". If no
+port is specified the value configured by `port` will be used.
 
 ===== port
 
-The port of the Redis server.
+Redis default port if hosts do not contain port number. The default is 6379.
```
Comment (ruflin) [inline libbeat/docs/outputconfig.asciidoc]: ... do not contain the port number ...

**Comment by ruflin** [comment]
@dedemorton Can you have a look at the doc changes?

File: libbeat/outputs/redis/client.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	minor, err = strconv.Atoi(string(versionRaw[2]))
+	if err != nil {
+		return nil, err
+	}
+
+	multiValue := major > 2 || (major == 2 && minor >= 4)
```
Comment (ruflin) [inline libbeat/outputs/redis/client.go]: A note would be nice why this differentiation is needed.

**Comment by ruflin** [comment]
LGTM. Would be good if perhaps @andrewkroh also could have a look.

File: libbeat/docs/outputconfig.asciidoc (hunk 1 above)
Comment (urso) [inline libbeat/docs/outputconfig.asciidoc]: Why not, it's default redis port

File: libbeat/outputs/redis/client.go (hunk 2 above)
Comment (urso) [inline libbeat/outputs/redis/client.go]: good point

Comment (urso) [inline libbeat/outputs/redis/client.go]: done

**Comment by ruflin** [comment]
@urso Can you squash the commits?

**Comment by andrewkroh** [comment]
Nice job. The code LGTM.

**Pull request merged**